### PR TITLE
Add upload certificate command to adam

### LINKED ISSUE #896: Add test covering change of controller signing certificate
Recently, we have seen many issues where EVE didn't properly handle a change in the controller signing certificate. It would be therefore beneficial to prepare an eden test in which the signing certificate used by Adam is changed after device is onboarded. This can be combined with poor network connectivity (which can be modeled with eden), etc. (conditions that triggered aforementioned issues).

However, first we must implement a handler in Adam for changing the signing certificate, see: https://github.com/lf-edge/adam/issues/118

CC @eriknordmark @rouming

### PR DESCRIPTION
**Pull request by europaul**
A new CLI option is added: on eden adam upload-cert the provided certificate is uploaded to <adam-ip>/admin/certs via an HTTP POST

NOTE: the new EDEN test needs to be run with an EVE version that includes [this fix](https://github.com/lf-edge/eve/pull/3675) and with the Adam version that includes [this](https://github.com/lf-edge/adam/pull/120) change.

Closes #896

### CODE REVIEW

**Comment by giggsoff** [comment]
It looks like the last part of the puzzle before moving to tests is to have some function, that will re-sign all config objects derived from the signing certificate [here](https://github.com/lf-edge/eden/blob/00d4d845276b04f38da8faa442f7fc8bc181ed26/pkg/expect/encrypt.go#L78), at least modify ControllerCertHash and ContextId, of CipherContexts correct?

**Comment by uncleDecart** [comment]
log_tests is failing, something is wrong with certificate path, looking into it...

**Comment by giggsoff** [comment]
Well, I am not sure will EVE-OS work fine with old certificate in CipherContexts and new certificate in Adam (attestation). I think we need some action to re-calculate old CipherContexts as we will potentially have a test with deploying of some app and check how it works after modification of certificate.

I see, but ControllerCertHash in CipherContext is about cert, not key.

**Comment by europaul** [comment]
@giggsoff I tried the following:
1. start eden and eve
2. `eden pod deploy -n nginx1 docker://nginx`
3. generate a new signing certificate from the same signing key (every parameter is the same except the validity period)
4. `eden adam upload-cert --cert-file ~/.eden/certs/signing-new.pem`
5. `eden pod deploy -n nginx2 docker://nginx`
6. check the logs and `eden pod ps` - both applications are running fine and I didn't see anything suspicious

I kept the old signing cert and key [here](https://github.com/lf-edge/eden/blob/00d4d845276b04f38da8faa442f7fc8bc181ed26/pkg/expect/encrypt.go#L78). Tbh I don't fully understand yet how the Expectation structures and package is used. Probably I'm missing something

**Comment by giggsoff** [comment]
Can you please check apps with any userdata (e.g. you made test with https://github.com/lf-edge/eden/pull/918). Inside of [expect](https://github.com/lf-edge/eden/blob/master/pkg/expect/encrypt.go) we try to encrypt two types of data: credentials of datastores and userdata.

**Comment by europaul** [comment]
@giggsoff I went on a roller-coaster ride through the code and found out that

we don't use the signing certificate to generate the symmetric key for the encryption, we only use the signing key
BUT
we use the hash of the signing certificate to generate IV for AES CFB, so this can mess the decryption up
BUT
we also send that IV with the cipher text in the cipher block, so decrypting should actually not be a problem...

**Comment by uncleDecart** [comment]
>But also I just thought maybe it's a good idea not to re-sign the old configs after changing the signing certificate - if EVE cannot deal with that this might just show a flaw in our design, since we the cloud might change the certificates frequently to ensure forward secrecy and we cannot expect it to re-sign all configs since the beginning of time. What do you think?

@europaul but that means if I have any previous certificate I can use it to sign configuration and send it, is it not? So potentially if I get any old certificate I can use it to change configuration, which sounds not secure. Or am I missing something?

**Comment by europaul** [comment]
@giggsoff you were right, I re-did the same test with applications with userdata and reboot like you suggested - after reboot EVE cannot decipher the userdata from the first application, because it is missing the old certificate that is required in the old cipherContext

File: pkg/openevec/adam.go
Comment (milan-zededa) [inline pkg/openevec/adam.go:139]: We also use cipher block here: https://github.com/lf-edge/eve-api/blob/main/proto/config/netconfig.proto#L171

Comment (europaul) [reply pkg/openevec/adam.go:139]: I noticed that we also use `CipherBlock` here, however this is not the version of EVE API that Eden uses. Why does Eden still use github.com/lf-edge/eve/api/go instead of github.com/lf-edge/eve-api/go?

Comment (giggsoff) [reply pkg/openevec/adam.go:139]: Fair question. We should update dependencies. But now we just rely on protobuf compatibility, so can use CipherBlock from [here](https://github.com/lf-edge/eve/blob/5a232aec4746e84cb9dc2b64f5d728c3926540df/api/go/config/netconfig.pb.go#L558). But it would be nice to support all of them (if it possible) in converter and write comment somewhere to not forget update in case of new cipherblocks.

Comment (uncleDecart) [reply pkg/openevec/adam.go:139]: IMO we should just move to eve-api/go repository. We should use same versions as EVE (i.e. we use EVE 11.04 as default EVE for Eden repo, which uses eve-api/go, so we need to use the same API)

File: pkg/controller/controller.go
Comment (uncleDecart) [inline pkg/controller/controller.go:20]: Should we implement this new Functions in every struct which implements Controller?

File: pkg/openevec/adam.go
Comment (milan-zededa) [reply pkg/openevec/adam.go:139]: I started with adam: https://github.com/lf-edge/adam/pull/121

Comment (europaul) [reply pkg/openevec/adam.go:139]: @milan-zededa thank you! I'm doing the update for Eden now

File: pkg/controller/controller.go
Comment (europaul) [reply pkg/controller/controller.go:20]: afaik Adam is the only one implementing Controller and it already implements this function... I think the code wouldn't compile if it wasn't implemented for all structs...

File: pkg/openevec/adam.go
Comment (milan-zededa) [reply pkg/openevec/adam.go:139]: @europaul I cherry-picked your commit to a [separate PR](https://github.com/lf-edge/eden/pull/935) so that we can merge it sooner (I need it to progress in my work). Hope it's OK

File: pkg/controller/functions.go
Comment (europaul) [reply pkg/controller/functions.go:169]: not really :thinking: we are still making the same request to the controller, but only returning the ECDH cert. so probably still waiting for all certs

File: pkg/expect/encrypt.go
Comment (europaul) [reply pkg/expect/encrypt.go:54]: as you can see later in the same function we use the first 16 bytes of the hash of `devCert` (which is the ECSH cert) to create `DeviceCertHash` in `cipherCtx` - this name seems to be consistent in the code I came across.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (europaul) [reply tests/eclient/testdata/ctrl_cert_change.txt]: This is embarrassing :facepalm: 

I wanted to put the certificate next to the the old signing cert, which is in `$HOME/.eden/certs`, so I created a new config variable `eden.cert-dir` for it now.

File: pkg/openevec/adam.go
Comment (europaul) [inline pkg/openevec/adam.go:118]: this code block is repeated for 4 different types of config here - they all have a `CipherData` field - but to put it in a function for reuse I would need to create an interface like the following in the eve-api repo. do you think it's suitable for that?
```go
type CipherDataHolder interface {
	GetCipherData() []byte
	SetCipherData(data []byte)
}
```

File: pkg/controller/functions.go
Comment (milan-zededa) [reply pkg/controller/functions.go]: OK

File: pkg/expect/encrypt.go
Comment (milan-zededa) [reply pkg/expect/encrypt.go]: OK

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (milan-zededa) [reply tests/eclient/testdata/ctrl_cert_change.txt]: I think it would be better to put the certificate under `eden.tests`. This is a temporary directory used by running tests.

File: pkg/openevec/adam.go
Comment (milan-zededa) [reply pkg/openevec/adam.go:118]: Yes, it would make sense to avoid repetition. But you can simply define the interface here in eden. Go code in eve-api repo is generated from proto and not meant for manual editing.

Comment (europaul) [reply pkg/openevec/adam.go:118]: True, but I would need to define the setter methods on those `config` structs and I can only do that local to the original definition

Comment (milan-zededa) [reply pkg/openevec/adam.go:118]: I see. Maybe some [reflect](https://pkg.go.dev/reflect) magic can do the job?

Comment (milan-zededa) [reply pkg/openevec/adam.go:118]: Btw. wouldn't Get method (which is available) enough? You can change the value of existing CipherData rather than to create a new instance and changing the pointer value.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (europaul) [reply tests/eclient/testdata/ctrl_cert_change.txt]: Done

File: pkg/openevec/adam.go
Comment (europaul) [reply pkg/openevec/adam.go]: You are right, I changed the value, but still had to jump through some hoops to avoid copying the lock that's part of every protobuf message

**Comment by milan-zededa** [comment]
@europaul LGTM but lets wait for the adam side to be merged: https://github.com/lf-edge/adam/pull/120
Then you will be able to update the [adam version](https://github.com/lf-edge/eden/blob/master/pkg/defaults/defaults.go#L57) and put your test into the test suite.
How long does the test take to execute? If it is fairly quick, we can put into the "Smoke tests" suite.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (giggsoff) [inline tests/eclient/testdata/ctrl_cert_change.txt:26]: Can you please clarify from where we expect the log? Are they aligned with latest changes of https://github.com/lf-edge/eden/blob/master/tests/eclient/image/files/entrypoint.sh?

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (giggsoff) [inline tests/eclient/testdata/ctrl_cert_change.txt:1]: It would be nice to write somewhere in comment here what exactly you check, some summary. E.g. validate re-encryption of app's userdata after controller certificate change with edge node reboot (with additional notes about 3 apps).
Additional question: can we somehow validate that we really change certificate from EVE-OS logs/info? I mean if we will put some kind of noop inside `eden adam upload-cert` (occasionally) looks like the test will pass.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (europaul) [reply tests/eclient/testdata/ctrl_cert_change.txt:26]: We expect this log message from [here](https://github.com/lf-edge/eve/blob/e5db3e506dc39135037d82d23c40785e069310cd/pkg/pillar/nireconciler/linux_reconciler.go#L400).

Sorry, I don't see how it relates to the file that you mentioned... But the branch has the newest version of `entrypoint.sh`

Comment (giggsoff) [reply tests/eclient/testdata/ctrl_cert_change.txt:29]: Ah. I see, it is log from EVE-OS (and you use `TestLog`). Seems `-app eclient1` not used [here](https://github.com/lf-edge/eden/blob/c2ba0752c0587dae3f5dd5d2fa74f56f6fcf2e8c/tests/lim/lim_test.go#L154-L193) and confuse me.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (europaul) [reply tests/eclient/testdata/ctrl_cert_change.txt:1]: Yes, we can check `/persist/certs/server-signing-cert.pem` in EVE. Added that to the test now together with the test description.

File: tests/eclient/testdata/ctrl_cert_change.txt
Comment (europaul) [reply tests/eclient/testdata/ctrl_cert_change.txt]: You are right, don't know where I got it from. Removed.

File: cmd/edenUtils.go
Comment (uncleDecart) [inline cmd/edenUtils.go:33]: how about newGenSigningCertCmd()?

File: pkg/controller/adam/adam.go
Comment (uncleDecart) [inline pkg/controller/adam/adam.go:203]: What does ECDH stands for?

File: pkg/utils/cipher.go
Comment (uncleDecart) [inline pkg/utils/cipher.go:102]: how about making this generic and putting this to `pkg/utils/slices.go`, i.e.

```golang

func CompareSlices[T comparable](lhs, rhs []T) bool {
  if len(lhs) != len(rhs) {
    return false
  }

  for i := range lhs {
    if lhs[i] != rhs[i] {
       return false
  }

   return true
}

```

**Review by uncleDecart — APPROVED** [review]

File: pkg/controller/adam/adam.go
Comment (europaul) [reply pkg/controller/adam/adam.go:203]: https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman

File: cmd/edenUtils.go
Comment (europaul) [reply cmd/edenUtils.go]: Done

File: pkg/utils/cipher.go
Comment (europaul) [reply pkg/utils/cipher.go]: Thanks, that's much better!

**Comment by europaul** [comment]
Adam used to get it's certificates and keys through envs. To unify the UX I switched this mechanism to getting paths to files with certificates and keys through the CLI parameters only. This way it doesn't require a POST API anymore.

Tell me how you like this solution!

P.S.: I will fix the commits later once we agree on the code.

**Comment by deitch** [comment]
> Adam used to get it's certificates and keys through envs. To unify the UX I switched this mechanism to getting paths to files with certificates and keys through the CLI parameters only. This way it doesn't require a POST API anymore.

To be clear, those path-based CLI flags already existed in Adam, correct? You are just taking advantage of them?

And Adam already reads those files with each request, rather than cache them in memory, so there is no need to tell Adam to reload the files, correct?

> This PR needs https://github.com/lf-edge/adam/pull/123 to work.

Does it? I think https://github.com/lf-edge/adam/pull/123 just gets rid of something that was confusing there. But this should not be blocked by it (unless I misunderstood)?

> Tell me how you like this solution!

Like!

**Comment by europaul** [comment]
> To be clear, those path-based CLI flags already existed in Adam, correct? You are just taking advantage of them?

> Does it? I think https://github.com/lf-edge/adam/pull/123 just gets rid of something that was confusing there. But this should not be blocked by it (unless I misunderstood)?

yes, the flags are already part of Adam, so no changes in Adam are required. The referenced commit is more of a clean-up. 

> And Adam already reads those files with each request, rather than cache them in memory, so there is no need to tell Adam to reload the files, correct?

yes, that is the case for the signing certificate which is relevant for this test

**Review by milan-zededa — APPROVED** [review]
LGTM, you can add the test into the "smoke" test suite: https://github.com/lf-edge/eden/blob/master/tests/workflow/smoke.tests.txt

**Comment by europaul** [comment]
added the test to the smoke test suite. ready to merge!

File: pkg/utils/cipher.go
Comment (giggsoff) [inline pkg/utils/cipher.go:94]: Curious, ill `bytes.Equal` work here?
Well, looks like the [logic](https://cs.opensource.google/go/go/+/release-branch.go1.20:src/bytes/bytes.go;drc=dcb90152a444be97fcc45bb12d176641c1b0d90e;l=20) inside of library function is suboptimal. So why to not use generics.

File: pkg/openevec/adam.go
Comment (giggsoff) [inline pkg/openevec/adam.go:62]: Looks like we can remove else after return.

File: pkg/expect/encrypt.go
Comment (giggsoff) [inline pkg/expect/encrypt.go:56]: Looks like a logic change, we mute error here before. As I understand we may have no certificate before attestation or for api v1, so please set it back to log and return nil error here. But return error instead of nil few lines below.

**Review by giggsoff — APPROVED** [review]
Great work! Thank you for your patience.
Few comments below.

File: pkg/utils/cryptoConfig.go
Comment (uncleDecart) [reply pkg/utils/cryptoConfig.go:81]: and then appendCipherCtx won't be needed at all, so it'll look something like,

```go
func AddCipherCtxToDev(dev *device.Ctx, cipherCtx *evecommon.CipherContext) {
	// check if we already have cipherCtx with the same certificates
	for _, c := range dev.GetCipherContexts() {
                sameCipherCtx := CompareSlices(c.DeviceCertHash, cipherCtx.DeviceCertHash) && CompareSlices(c.ControllerCertHash, cipherCtx.ControllerCertHash)
		if sameCipherCtx {
			cipherCtx = c
                        return
		}
	}
       
       dev.SetCipherContexts(append(dev.GetCipherContexts(), cipherCtx))
}
```

File: pkg/openevec/adam.go
Comment (uncleDecart) [inline pkg/openevec/adam.go:60]: Do we need 0644 permissions here or 0600 will suffice?

**Comment by uncleDecart** [comment]
once https://github.com/lf-edge/adam/pull/120 is merged and we pull latest adam we can merge this

File: pkg/openevec/adam.go
Comment (uncleDecart) [reply pkg/openevec/adam.go:60]: Okay, I'm fine with that

File: pkg/utils/cryptoConfig.go
Comment (europaul) [reply pkg/utils/cryptoConfig.go:81]: Absolutely! I think it would be even cleaner to return the cipher context that was added or the matching existing one, since we don't compare all the fields and there might potentially be discrepancies

File: pkg/expect/encrypt.go
Comment (europaul) [reply pkg/expect/encrypt.go:56]: I agree about the ECDH certificate, but what do you mean by "few lines below? if it's about signing cert, then it should be the same as for the ECDH cert, e.g. api v1 - no error

Comment (giggsoff) [reply pkg/expect/encrypt.go:56]: Ah, you are right, signing key is also come from APIv2. So please log and return here. Nothing required for lines below.

File: pkg/utils/x509.go
Comment (europaul) [reply pkg/utils/x509.go]: Done.

**Comment by europaul** [comment]
@giggsoff @uncleDecart thanks for the great comments!

> once https://github.com/lf-edge/adam/pull/120 is merged and we pull latest adam we can merge this

https://github.com/lf-edge/adam/pull/120 was superseded by https://github.com/lf-edge/adam/pull/123, so I'm gonna close it. However the current PR doesn't require any changes in Adam and can be merged independently

**Comment by deitch** [comment]
Some yetus complains, and some eden tests failing. Are those failures because of this, because of the other changes, or transient?

**Comment by europaul** [comment]
> Some yetus complains, and some eden tests failing. Are those failures because of this, because of the other changes, or transient?

I'd say they fail, because Eden workflows still run with an older EVE version. [This PR](https://github.com/lf-edge/eden/pull/942) is here to fix it

**Comment by uncleDecart** [comment]
@europaul can you rebase to master so that we can rerun tests?

**Review by uncleDecart — APPROVED** [review]
Running tests

**Comment by uncleDecart** [comment]
There's a failure on rebooting device on ZFS filesystem. Problem is persistent and also occurred before, need to investigate zfs failure before merging this

**Pull request merged**
